### ISSUE TITLE
Some kind of upgrade solution

### ISSUE THREAD

**Issue by loidor** [Libki/libki-server#41]
Upgrading isn't simply to git pull and run the installer again, since there's some new perl packages and whatnot depending on what version you're coming from.

In some distant future there might even be an upgrade button in the admin GUI, but at the point being that's not it.

So for now, I'll write up a wiki page (see #40) on how to reinstall the server while keeping the database.

I'm posting this (and a couple of other issues) so I don't forget about them.

**Comment by kylemhall** [comment]
You are correct, but it shouldn't be too much harder.
`cpanm  --installdeps .
perl installer/update_db.pl`
Should do it for the most part.
Really, I imagine each release should have upgrade notes if need be.

**Comment by loidor** [comment]
Yes, that sounds like a good way to go at it. I'm glad you thought about releases - this will make very many things a whole lot easier in administrating ways - easier to spot what went wrong where, being dead-sure that it works on a specific platform, et cetera.

**Comment by kylemhall** [comment]
What do you think of having Libki auto-update the db schema instead of needing to run the installer manually?

**Comment by loidor** [comment]
As long as we set a standard for what's needed, that shouldn't be a problem. I.e. the new auto-deletion won't work in default Jessie since datetime current_timestamp wasn't implemented until mysql 5.6.5. 

However, I should really start digging into getting it all up and running on Stretch and the new Ubuntu server, so it should be future-safe for another five years or something.

**Comment by loidor** [comment]
I'll be looking in to this next.

I'll write something that makes a backup, plain and simple. That will be needed when the whole server needs to be upgraded.

I'll also add an importer that takes whatever backup you want and stuffs it in to the new database.

There will also be a page in the manual about this, where I can go into more detail on things like backups via       cron, how to send them to an ftp server, how to clean out a database that needs restoring, how to copy your brand new backup to a thumb drive (since people will need to upgrade from Jessie) and so on.

**Comment by loidor** [comment]
I ran in to a weird problem. Should be an easy fix, but I'm curious on why it happens.

I populate the database via ```update_db.pl```. I can create new users just fine with ```create_user.pl```.

I drop the database and create it again, but this time i populate it from a mysqldump. The users created comes in nicely, but suddenly I can't create new users. ```'notes'``` is giving me problems. Mysql is mad because "Field 'notes' doesn't have a default value". And, well, notes is text, so it can't have a default value in current mysql versions, and it's set to NOT NULL, and if nothing is passed... Yeah. 

But why does it work in the first place?

Anyhow I'll add a quick patch allowing NULL in notes.

**Comment by kylemhall** [comment]
Very odd! I've merged your pull request iirc!

**Comment by loidor** [comment]
I've been thinking more about this, and I think the best way would be to add an `update` flag to the installer, that runs ` service libki stop && libki-backup && git pull && cpanm -n --installdeps . && perl installer/update_db.pl && service libki start`. 

Everybody doesn't want automatic upgrades (it it ain't broke...), and this way it's still really easy to do. And it makes a backup of the database, just in case.

**Comment by kylemhall** [comment]
Sounds good!

**Comment by masterfink** [comment]
Where is the database name defined in installer/update_db.pl? I have a server with 3 instances of libki on it, but can't seem to run that script on any database other than the default "libki".

**Comment by loidor** [comment]
It sources the db info from libki_local.conf. In the top part you have:
``` 
<Model::DB>
    <connect_info>
        dsn   dbi:mysql:libki
        user   libki
        password "PASSWORD"
        auto_savepoint 1
        quote_names 1
    </connect_info>
</Model::DB>
```

The `dsn` part is your database connection. If you want to change it, this is its format:

`dbi:SQL Platform:database_name:host_name:port`

**Comment by loidor** [comment]
I finally had some time to work on this, and have included an upgrade/rollback feature in the installer. I'm working on a separate patch to make it possible to run it from older versions as well, and once that's done I'll make a PR.

**Comment by masterfink** [comment]
Thanks Ioidor for your hard work! Has made the migration of Libki much easier.

**Comment by loidor** [comment]
After testing this thoroughly I'm sad to say I will need to rewrite most of it in order to get it to work with older versions. There's simply too much that have changed compared to installs from, say, early 2017 (I know of a few of these). 

The upgrade will make a dump of the database, some service data and libki_local.conf, then wipe everything libki related, reinstall it all, modify the db dump so it can be safely imported and modify the new service and conf and it should be properly up and running. Sure, it's way more complex, but it will be invaluable for those libraries with old installs. Also, it's fun :D

With that said, I'll also combine it with some sort of import feature for those in need to upgrade their servers from Debian Jessie to Stretch or Ubuntu 18.04.

**Comment by loidor** [comment]
Included in #114
